File: Lab3/v4_final/Python/player.py
# player.py - Игрок с управлением и стрельбой

import pygame
import math
from config import *

class Player:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.hull_angle = 0
        self.size = PLAYER_SIZE
        self.base_speed = PLAYER_BASE_SPEED
        self.health = PLAYER_MAX_HEALTH
        self.max_health = PLAYER_MAX_HEALTH
        self.shoot_cooldown = 0
        self.score = 0
        self.radius = COLLISION_RADIUS_PLAYER
        
        self._load_image()
    
    def _load_image(self):
        """Загрузка спрайта игрока"""
        try:
            self.image = pygame.image.load('img/player/player_up.png').convert_alpha()
            self.image = pygame.transform.scale(self.image, (self.size, self.size))
        except pygame.error:
            self.image = self._create_fallback_image()
    
    def _create_fallback_image(self):
        """Создание заглушки если спрайт не найден"""
        surf = pygame.Surface((self.size, self.size), pygame.SRCALPHA)
        pygame.draw.polygon(surf, (0, 100, 255), [
            (self.size//2, 0),
            (self.size, self.size),
            (self.size//2, self.size*0.7),
            (0, self.size)
        ])
        return surf
    
    def update(self, keys, obstacles):
        """Обновление состояния игрока"""
        self._handle_rotation(keys)
        self._handle_movement(obstacles)
        self._update_cooldown()
    
    def _handle_rotation(self, keys):
        """Обработка поворота корпуса"""
        if keys[pygame.K_a] and self.hull_angle > -PLAYER_MAX_ANGLE:
            self.hull_angle -= PLAYER_ROTATION_SPEED
        elif keys[pygame.K_d] and self.hull_angle < PLAYER_MAX_ANGLE:
            self.hull_angle += PLAYER_ROTATION_SPEED
        else:
            # Автовозврат к прямому курсу
            if abs(self.hull_angle) < PLAYER_AUTO_RETURN_SPEED:
                self.hull_angle = 0
            elif self.hull_angle > 0:
                self.hull_angle -= PLAYER_AUTO_RETURN_SPEED
            else:
                self.hull_angle += PLAYER_AUTO_RETURN_SPEED
        
        self.hull_angle = max(-PLAYER_MAX_ANGLE, min(PLAYER_MAX_ANGLE, self.hull_angle))
    
    def _handle_movement(self, obstacles):
        """Обработка движения и коллизий"""
        old_x, old_y = self.x, self.y
        
        # Движение вперёд и боковое смещение
        self.y -= self.base_speed
        side_speed = (self.hull_angle / PLAYER_MAX_ANGLE) * PLAYER_SIDE_SPEED_MULTIPLIER
        self.x += side_speed
        
        # Ограничение краёв экрана
        self.x = max(PLAYER_EDGE_MARGIN, min(SCREEN_WIDTH - PLAYER_EDGE_MARGIN, self.x))
        
        # Проверка коллизий
        if self._check_collisions(obstacles):
            self._handle_collision(old_x, old_y)
    
    def _check_collisions(self, obstacles):
        """Проверка столкновений с препятствиями"""
        for obstacle in obstacles:
            # Острова
            if hasattr(obstacle, 'collides_with'):
                if obstacle.collides_with(self.x, self.y, self.radius):
                    return True
            # Берега
            elif hasattr(obstacle, 'contains_point'):
                if obstacle.contains_point(self.x, self.y, self.radius):
                    return True
        return False
    
    def _handle_collision(self, old_x, old_y):
        """Обработка столкновения"""
        self.health -= PLAYER_COLLISION_DAMAGE
        self.x, self.y = old_x, old_y
        self.y += PLAYER_COLLISION_PUSHBACK
    
    def _update_cooldown(self):
        """Обновление таймера перезарядки"""
        if self.shoot_cooldown > 0:
            self.shoot_cooldown -= 1
    
    def collides_with(self, x, y, radius):
        """Проверка столкновения с точкой"""
        dx = self.x - x
        dy = self.y - y
        distance = math.sqrt(dx*dx + dy*dy)
        return distance < self.radius + radius
    
    def shoot(self):
        """Стрельба с учетом угла поворота"""
        if self.shoot_cooldown > 0:
            return []
        
        self.shoot_cooldown = PLAYER_SHOOT_COOLDOWN
        
        # Базовый угол - вверх
        angle = math.radians(-90)
        
        # Корректировка угла в зависимости от поворота
        if self.hull_angle > PLAYER_MIN_ANGLE_FOR_SIDE_SHOT:
            angle += math.radians(PLAYER_SHOOT_ANGLE_OFFSET)
        elif self.hull_angle < -PLAYER_MIN_ANGLE_FOR_SIDE_SHOT:
            angle -= math.radians(PLAYER_SHOOT_ANGLE_OFFSET)
        
        from projectile import Projectile
        return [Projectile(self.x, self.y, angle)]
    
    def take_damage(self, amount):
        """Получение урона"""
        self.health -= amount
        return self.health <= 0
    
    def draw(self, screen, camera_y):
        """Отрисовка игрока"""
        y_screen = int(self.y - camera_y)
        rotated = pygame.transform.rotate(self.image, -self.hull_angle)
        rect = rotated.get_rect(center=(int(self.x), y_screen))
        screen.blit(rotated, rect.topleft)

File: Lab4_1/flutter_joystick_app/lib/game/models/player.dart
import 'dart:math';
import '../game_config.dart';
import 'projectile.dart';
import 'island.dart';

class Player {
  double x;
  double y;
  late double hullAngle;  // угол поворота корпуса в градусах
  late double size;
  late double baseSpeed;
  late double health;
  late double maxHealth;
  late int shootCooldown;
  late int score;
  late double radius;
  late double totalDistance;  // Общее пройденное расстояние в пикселях

  Player(this.x, this.y) {
    hullAngle = 0;
    size = GameConfig.playerSize;
    baseSpeed = GameConfig.playerBaseSpeed;
    health = GameConfig.playerMaxHealth;
    maxHealth = GameConfig.playerMaxHealth;
    shootCooldown = 0;
    score = 0;
    radius = GameConfig.collisionRadiusPlayer;
    totalDistance = 0.0;
  }

  void update(double joystickX, double joystickY, List<dynamic> obstacles) {
    // Управление через джойстик
    // joystickX и joystickY от -100 до 100
    // joystickX управляет поворотом и боковым движением (чем больше отклонение, тем быстрее поворот)
    // joystickY управляет скоростью вперёд (чем больше отклонение, тем быстрее)
    
    _handleRotation(joystickX);
    _handleMovement(joystickX, joystickY, obstacles);
    _updateCooldown();
  }

  void _handleRotation(double joystickX) {
    // Джойстик X управляет углом поворота
    // Чем больше отклонение, тем быстрее поворот
    double rotationSpeed = GameConfig.playerRotationSpeed * (joystickX.abs() / 100.0).clamp(0.1, 1.0);
    double targetAngle = (joystickX / 100.0) * GameConfig.playerMaxAngle;
    
    // Плавный поворот
    double angleDiff = targetAngle - hullAngle;
    if (angleDiff.abs() > rotationSpeed) {
      hullAngle += angleDiff.sign * rotationSpeed;
    } else {
      hullAngle = targetAngle;
    }
    
    // Автовозврат к прямому курсу если джойстик в центре
    if (joystickX.abs() < 5) {
      if (hullAngle.abs() < GameConfig.playerAutoReturnSpeed) {
        hullAngle = 0;
      } else {
        hullAngle -= hullAngle.sign * GameConfig.playerAutoReturnSpeed;
      }
    }
    
    hullAngle = hullAngle.clamp(-GameConfig.playerMaxAngle, GameConfig.playerMaxAngle);
  }

  void _handleMovement(double joystickX, double joystickY, List<dynamic> obstacles) {
    double oldX = x;
    double oldY = y;
    
    // Скорость зависит от отклонения джойстика Y (чем больше отклонение, тем быстрее)
    // joystickY: положительное = вперёд (вверх), отрицательное = назад (вниз)
    // Но в игре всегда движемся вперёд, просто меняется скорость
    double forwardSpeedMultiplier = 1.0 + (joystickY.abs() / 100.0) * 0.5;
    double forwardSpeed = baseSpeed * forwardSpeedMultiplier;
    
    // Движение вперёд (вверх по экрану в игре) - всегда
    y -= forwardSpeed;
    
    // Накапливаем пройденное расстояние с учетом реальной скорости
    totalDistance += forwardSpeed;
    
    // Боковое смещение от угла поворота И от джойстика X
    // Чем больше отклонение X, тем больше боковое смещение
    double sideSpeedFromAngle = (hullAngle / GameConfig.playerMaxAngle) * GameConfig.playerSideSpeedMultiplier;
    double sideSpeedFromJoystick = (joystickX / 100.0) * 2.0;  // Прямое управление боковым движением
    x += sideSpeedFromAngle + sideSpeedFromJoystick;
    
    // Ограничение краёв экрана
    x = x.clamp(GameConfig.playerEdgeMargin, GameConfig.screenWidth - GameConfig.playerEdgeMargin);
    
    // Проверка коллизий
    if (_checkCollisions(obstacles)) {
      _handleCollision(oldX, oldY);
    }
  }

  bool _checkCollisions(List<dynamic> obstacles) {
    for (var obstacle in obstacles) {
      if (obstacle is Island) {
        if (obstacle.collidesWith(x, y, radius)) {
          return true;
        }
      } else if (obstacle is Shore) {
        if (obstacle.collidesWith(x, y, radius)) {
          return true;
        }
      }
    }
    return false;
  }

  void _handleCollision(double oldX, double oldY) {
    health -= GameConfig.playerCollisionDamage;
    x = oldX;
    y = oldY;
    y += GameConfig.playerCollisionPushback;
  }

  void _updateCooldown() {
    if (shootCooldown > 0) {
      shootCooldown--;
    }
  }

  bool collidesWith(double checkX, double checkY, double checkRadius) {
    double dx = x - checkX;
    double dy = y - checkY;
    double distance = sqrt(dx * dx + dy * dy);
    return distance < radius + checkRadius;
  }

  List<Projectile>? shoot() {
    if (shootCooldown > 0) {
      return null;
    }
    
    shootCooldown = GameConfig.playerShootCooldown;
    
    double angle = -pi / 2;  // Вверх
    
    if (hullAngle > GameConfig.playerMinAngleForSideShot) {
      angle += GameConfig.playerShootAngleOffset * pi / 180;
    } else if (hullAngle < -GameConfig.playerMinAngleForSideShot) {
      angle -= GameConfig.playerShootAngleOffset * pi / 180;
    }
    
    return [Projectile(x, y, angle, true)];
  }

  bool takeDamage(double amount) {
    health -= amount;
    return health <= 0;
  }
}

